File: src/Runtime/Sentence/SentenceImport.cpp
#include "SentenceImport.h"
#include "../Module.h"
#include "../ModulePool.h"

using namespace peak::interpreter;

SentenceImport::SentenceImport(const std::string& moduleName)
	: _moduleName(moduleName) {
}
ExecuteResult SentenceImport::Execute(std::shared_ptr<Space> space) {
	if (space->GetSpaceType() != SpaceType::None) {
		ErrorLogger::LogRuntimeError(_moduleName);
		ErrorLogger::LogRuntimeError(ErrorRuntimeCode::Import, "The module \"" + _moduleName + "\" can't import here!");
		return ExecuteResult::Failed;
	}
	auto module = ModulePool::GetInstance()->UseModule(_moduleName);
	if (!module) {
		ErrorLogger::LogRuntimeError(_moduleName);
		ErrorLogger::LogRuntimeError(ErrorRuntimeCode::Import, "The module \"" + _moduleName + "\" not found!");
		return ExecuteResult::Failed;
	}
	if (!space->UseModule(module)) {
		ErrorLogger::LogRuntimeError(_moduleName);
		ErrorLogger::LogRuntimeError(ErrorRuntimeCode::Import, "The module \"" + _moduleName + "\" already exists!");
		return ExecuteResult::Failed;
	}
	return ExecuteResult::Successed;
}

File: src/Runtime/Sentence/SentenceImport.cs
/*
	PeakScript: https://github.com/Jenocn/PeakScript
	By Jenocn: https://jenocn.github.io
*/

namespace peak.interpreter {

	public class SentenceImport : Sentence {
		private string _moduleName = "";
		public SentenceImport(string moduleName) {
			_moduleName = moduleName;
		}
		public override ExecuteResult Execute(Space space) {
			if (space.spaceType != SpaceType.None) {
				ErrorLogger.LogRuntimeError(_moduleName);
				ErrorLogger.LogRuntimeError(ErrorRuntimeCode.Import, "The module \"" + _moduleName + "\" can't import here!");
				return ExecuteResult.Failed;
			}
			var module = ModulePool.instance.UseModule(_moduleName);
			if (module == null) {
				ErrorLogger.LogRuntimeError(_moduleName);
				ErrorLogger.LogRuntimeError(ErrorRuntimeCode.Import, "The module \"" + _moduleName + "\" not found!");
				return ExecuteResult.Failed;
			}
			if (!space.UseModule(module)) {
				ErrorLogger.LogRuntimeError(_moduleName);
				ErrorLogger.LogRuntimeError(ErrorRuntimeCode.Import, "The module \"" + _moduleName + "\" already exists!");
				return ExecuteResult.Failed;
			}
			return ExecuteResult.Successed;
		}

	}

} // namespace peak.interpreter